### ISSUE TITLE
Cannot import csv after export

### ISSUE THREAD

**Issue by shadow00** [vmiklos/plees-tracker#1]
I wanted to edit the start/stop timestamps of the last sleep, so I exported the csv file and manually changed it with an editor.  
Now I'm trying to import the file again, but the import file selector does not let me choose any file from internal memory - all the files are greyed out and cannot be selected.

**Comment by vmiklos** [comment]
Indeed. This used to work.. now all I get is `java.io.IOException: Stream Closed`. I'll see what's going on here, thanks for reporting.

**Comment by vmiklos** [comment]
This now works for me. How did you install plees tracker, fdroid? If so, I should probably cut a release. :-)

**Comment by shadow00** [comment]
Yes, I downloaded it from f-droid. I'll try it as soon as you release it!

**Comment by vmiklos** [comment]
v4.0 is out now, ETA for the fdroid build is a ~week.

**Comment by vmiklos** [comment]
<https://gitlab.com/fdroid/fdroiddata/commit/8b55e5b87daaaf974d76faedab2b38abdcfbfcf1> picked up the release, now waiting for a build to appear at <https://f-droid.org/repo/hu.vmiklos.plees_tracker_4.apk>.

**Comment by vmiklos** [comment]
Now even the apk is there: https://f-droid.org/repo/hu.vmiklos.plees_tracker_4.apk, one week as ETA wasn't that off. :-)

**Comment by shadow00** [comment]
So, a few days ago I tried importing a csv again after updating to v4.0, and once again I couldn't open any file from the system's file selector. Everything was still greyed out.
And I now just realized why: your app does not have the system permission to access files in memory! In fact, it doesn't even request that permission from the system.

However, I was still able to import a file from an external file manager (I'm using Simple File Manager), which does have memory access.

**Comment by shadow00** [comment]
Also, would it be possible to have a way to edit each recorded entry? Sometimes I forget to press start before falling asleep, and manually editing unix timestamps is kind of a pain. Maybe tapping/long pressing an entry could open a date and time editor (like google calendar's, for example) and allow the user to change the start and stop time and date.

**Comment by vmiklos** [comment]
> And I now just realized why: your app does not have the system permission to access files in memory! In fact, it doesn't even request that permission from the system.

Oh, this is interesting. Do you rename the file between the export and import? You're right, reading https://github.com/vmiklos/plees-tracker/blob/862fdd74e542a62035fa5920bdb82baf5a8c598e/app/src/main/java/hu/vmiklos/plees_tracker/DataModel.kt#L92 shows that export calls takePersistableUriPermission(), while import does not. I always tested import with a file that was previously exported. ;-)

> Also, would it be possible to have a way to edit each recorded entry?

Sure, why not. But please open a separate issue for that.

**Comment by shadow00** [comment]
> I always tested import with a file that was previously exported.
Me too: I exported the file, changed an entry, then tried to import it back.

> But please open a separate issue for that.
Ok, I'll do that right away. Thank you!

**Comment by vmiklos** [comment]
OK, then I need more info: how did you change it? What I tried is:

1) Open plees, menu -> export -> accept to put plees-tracker.csv to Downloads.
2) Open the Files app, go to Downloads, long push on plees-tracker.csv, "..." menu, copy to, select "Copy" to create "plees-tracker (1).csv".
3) Go back to plees, menu -> import -> select "plees-tracker (1).csv" -> "import finished successfully".

Does this fail for you?

**Comment by shadow00** [comment]
Yes, it does. Everything is still greyed out. I can only import a file if I open the toast menu on the top left and use my file manager to select the file.
![Screenshot 1](https://i.imgur.com/yRuYoQA.jpg)
![Screenshot 2](https://i.imgur.com/WpGxgiX.jpg)

**Comment by vmiklos** [comment]
Okay, I don't have an idea about this additional problem at the moment, glad that you have a workaround ;-)

**Comment by SebiderSushi** [comment]
As it seems this is an issue with the mime type:

https://github.com/vmiklos/plees-tracker/blob/508f8caefd358e65276a534ba2eabf4d7ea8e9ac/app/src/main/java/hu/vmiklos/plees_tracker/MainActivity.kt#L160

First, [i figured out](https://stackoverflow.com/a/28172086) that if you specify `text/*` as the mime type, the csv file can be selected and imported (naturally other text files as well).

Then i found that while importing, `ContentResolver.getType()` returned `text/comma-separated-values` for the uri passed onto `ViewModel.importData()`.  
So in lights of [another answer]( https://stackoverflow.com/a/43634560) from the stackoverflow link, [this](https://github.com/SebiderSushi/plees-tracker/commit/b4acd95aa31768696a8483231b622de5dcadb6bb) is how it works as expected on my device (android 9).

However, if data integrity is ensured (i.e. the parser can be expected to not trip over other text files), i think just specifiying the mimeType as `text/*` can be useful as well, in case an android device has yet another way of specifying the mime type or simply misinterprets it.
In my opinion, a user can be endowed with the "responsibility" of selecting the right file and might even want to do it.

**Comment by vmiklos** [comment]
Interesting. I just tested this manually, master works here just fine. Though I'm on Android 10. I also wrote an androidTest for this back then (see testImportExport() in app/src/androidTest/java/hu/vmiklos/plees_tracker/MainActivityInstrumentedTest.kt), and Travis intentionally runs on Android 5.1 (API level 22).

Anyhow, perhaps the test is poor and there is still to be fixed here. :-) PRs are welcome, as long as the already working cases keep working.

**Comment by SebiderSushi** [comment]
I don't think this can properly be tested as the issue is with the behavior of the files picker of the android system, and thus external to `plees-tracker`. I can't think of any proper test in this situation.
`plees-tracker` isn't where there are problems with the mime type handling, it is just the android file picker who does not accept selecting the proper file because it represents the mimeType differently.

I suggest to merge my commit, or decide that all plain text files are acceptable for the parser and just change the `intent.type` to `text/*`.

**Comment by vmiklos** [comment]
Fine, could you please submit a pull request with your commit in it?

If it clearly states this is a workaround for non-Google Android flavors (or old Android versions, etc), I don't mind :-)

**Comment by SebiderSushi** [comment]
Well the thing is i have no clue where this is coming from or whether it could be called "non-Google".
I have absolutely no background info. All i did was capture the uri passed into `ViewModel.importData()` and find out what was returned by `ContentResolver.getType()`.
I tested this on LineageOS 15.1 (oreo) and 16.0 (pie) and i doubt very much that they'd change such a thing.

**TL;DR: I cannot say what exactly this workaround is for, it could be researched but honestly - what could be gained from it?**

**Comment by vmiklos** [comment]
Thanks, I'll take a look.
